piecedirectory/piecedirectory.go

```diff
@@ -406,119 +406,140 @@ func (cr *countingReader) Read(p []byte) (n int, err error) {
-	}
-
-	chunkSize := len(indexData.Entries) / concurrency
-	results := make([][]datasegment.SegmentDesc, concurrency)
-
-	var eg errgroup.Group
-	for i := 0; i < concurrency; i++ {
-		i := i
-		eg.Go(func() error {
-			start := i * chunkSize
-			end := start + chunkSize
-			if i == concurrency-1 {
-				end = len(indexData.Entries)
-			}
-
-			res, err := validateEntries(indexData.Entries[start:end])
-			if err != nil {
-				return err
-			}
-
-			results[i] = res
-
-			return nil
-		})
-	}
-
-	if err := eg.Wait(); err != nil {
-		return nil, fmt.Errorf("could not calculate valid entries: %w", err)
-	}
-
-	validSegments := make([]datasegment.SegmentDesc, 0, len(indexData.Entries))
-	for _, res := range results {
-		validSegments = append(validSegments, res...)
-	}
-
-	if len(validSegments) == 0 {
-		return nil, fmt.Errorf("no data segments found")
-	}
-
-	log.Debugw("podsi: validated data segment index", "validSegments", len(validSegments), "time", time.Since(start).String())
-	start = time.Now()
-	readsCnt = 0
-
-	recs := make([]model.Record, 0)
-	for _, s := range validSegments {
-		segOffset := s.UnpaddedOffest()
-		segSize := s.UnpaddedLength()
-
-		lr := io.NewSectionReader(r, int64(segOffset), int64(segSize))
-
-		cr = &countingReader{
-			Reader: lr,
-			cnt:    &readsCnt,
-		}
-
-		subRecs, err := parseRecordsFromCar(bufio.NewReaderSize(cr, PodsiBuffesrSize))
-		if err != nil {
-			// revisit when non-car files supported: one corrupt segment shouldn't translate into an error in other segments.
-			return nil, fmt.Errorf("could not parse data segment #%d at offset %d: %w", len(recs), segOffset, err)
-		}
-		for i := range subRecs {
-			subRecs[i].Offset += segOffset
-		}
-		recs = append(recs, subRecs...)
-	}
-
-	log.Debugw("podsi: parsed records from data segments", "recs", len(recs), "reads", readsCnt, "time", time.Since(start).String())
-
-	return recs, nil
+    concurrency := runtime.NumCPU()
+    if concurrency < PodsiMinConcurrency {
+        concurrency = PodsiMinConcurrency
+    }
+    if concurrency > PodsiMaxConcurrency {
+        concurrency = PodsiMaxConcurrency
+    }
+
+    log.Debugw("podsi: ", "bufferSize", PodsiBuffesrSize, "validationConcurrency", concurrency)
+    start := time.Now()
+
+    ps := abi.UnpaddedPieceSize(unpaddedSize).Padded()
+    dsis := datasegment.DataSegmentIndexStartOffset(ps)
+    if _, err := r.Seek(int64(dsis), io.SeekStart); err != nil {
+        log.Debugw("Failed to seek to data segment index", "error", err)
+        return nil, fmt.Errorf("could not seek to data segment index: %w", err)
+    }
+
+    // Debug: Print the file size
+    file, ok := r.(io.ReadSeeker)
+    if ok {
+        fileSize, err := file.Seek(0, io.SeekEnd)
+        if err != nil {
+            log.Debugw("Failed to get file size", "error", err)
+        } else {
+            log.Debugw("File size", "size", fileSize)
+        }
+        _, err = file.Seek(0, io.SeekStart) // Return the file pointer to the start
```

Comment:
Shouldn't we set the seeker back to the start of the segment index offset same as line 422? The current code would work for 32 GiB files but might be problematic for smaller deals.

Reply:
yes, `dsis` is kind of important for the `parseDataSegmentIndex` usage below, `0` is not a good idea here